lib/crontab/date_helper.ex

```diff
@@ -289,25 +291,59 @@ defmodule Crontab.DateHelper do
     if Date.day_of_week(date) == weekday do
       day
     else
-      last_weekday(add(date, -1, :day), weekday, :end)
+      last_weekday(shift(date, -1, :day), weekday, :end)
     end
   end
 
   @doc false
-  def add(datetime = %NaiveDateTime{}, amt, unit), do: NaiveDateTime.add(datetime, amt, unit)
+  @spec shift(date, integer, unit, ambiguity_opts) :: date
+  def shift(datetime, amt, unit, ambiguity_opts \\ [:later])
 
-  def add(datetime = %DateTime{}, amt, unit) do
+  def shift(datetime = %NaiveDateTime{}, amt, unit, _), do: NaiveDateTime.add(datetime, amt, unit)
+
+  def shift(datetime = %DateTime{}, amt, unit, _) when unit in [:second, :minute] do
+    DateTime.add(datetime, amt, unit)
+  end
+
+  def shift(datetime = %DateTime{}, amt, unit, ambiguity_opts) do
     candidate = DateTime.add(datetime, amt, unit)
-    adjustment = datetime.std_offset - candidate.std_offset
-    adjusted = DateTime.add(candidate, adjustment, :second)
 
-    if adjusted.std_offset != candidate.std_offset do
-      candidate
+    case DateTime.from_naive(DateTime.to_naive(candidate), candidate.time_zone) do
+      {:ambiguous, earlier, later} ->
+        resolve_ambiguity(datetime, earlier, later, amt, ambiguity_opts)
+
+      _ ->
+        resolve_potential_gap(datetime, candidate, amt, unit)
+    end
+  end
+
+  def resolve_ambiguity(from_time, earlier, later, amt, ambiguity_opts) do
+    if (:earlier in ambiguity_opts and from_time < earlier) or amt < 0 do
+      earlier
     else
-      case DateTime.from_naive(DateTime.to_naive(adjusted), adjusted.time_zone) do
-        {:ambiguous, _, target} -> target
-        {:ok, target} -> target
-      end
+      later
+    end
+  end
+
+  def resolve_potential_gap(%{std_offset: n}, candidate = %{std_offset: n}, _, _), do: candidate
+
+  def resolve_potential_gap(_, candidate, _, unit) when unit in [:second, :minute, :hour] do
+    candidate
+  end
+
+  def resolve_potential_gap(%{std_offset: n}, candidate = %{std_offset: m}, amt, _) do
+    cond do
```

Comment:
I think this needs a bit more detail since DST shifts are not guaranteed to be one hour.

For example, Lord Howe Island is UTC +10:30 / +11:00